src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -2383,19 +2448,33 @@ internal void ReadSni(TaskCompletionSource<object> completion)
             {
                 Debug.Assert(completion != null, "Async on but null asyncResult passed");
 
-                if (_networkPacketTimeout == null)
+                // if the state is currently stopped then change it to running and allocate a new identity value from 
+                // the identity source. The identity value is used to correlate timer callback events to the currently
+                // running timeout and prevents a late timer callback affecting a result it does not relate to
+                int previousTimeoutState = Interlocked.CompareExchange(ref _timeoutState, TimeoutRunning, TimeoutStopped);
+                if (previousTimeoutState == TimeoutStopped)
                 {
-                    _networkPacketTimeout = ADP.UnsafeCreateTimer(
-                        new TimerCallback(OnTimeout),
-                        null,
-                        Timeout.Infinite,
-                        Timeout.Infinite);
+                    Debug.Assert(_timeoutIdentityValue == 0, "timer was previously stopped without resetting the _identityValue");
+                    _timeoutIdentityValue = Interlocked.Increment(ref _timeoutIdentitySource);
```

Comment:
As long as the value differs from the previous value it doesn't matter about the value itself. It's never compared with 0 so wrapping around poses no problem. I considered this and the possibility of using a long but instead went with the int.

Reply:
I just learned something new about Interlocked.Increment: "This method handles an overflow condition by wrapping: if location = Int32.MaxValue, location + 1 = Int32.MinValue. **No exception is thrown.**" I assumed it threw. 😆

Reply:
Even simple ++ would have worked by default since the runtime defaults to unchecked mode for speed. We'd have to force checked mode to get an exception. The interlocked here isn't strictly required but it makes it safe if multiple callers end up here at the same time which shouldn't be possible but i've learnt that "shouldn't be possible" happens too often.